### PR TITLE
Document all C# API pubic interfaces

### PR DESCRIPTION
**Pull request by yuslepukhin**
**Description**:
Supply missing documentation pieces. Prepare to generate doc pages.

### CODE REVIEW

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/Microsoft.ML.OnnxRuntime.Tests.csproj [hunk 1]
```diff
@@ -60,12 +60,15 @@
   </ItemGroup> 
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.NETCore.Targets" Version="2.1.0" />
-    <PackageReference Include="System.Runtime.CompilerServices.Unsafe" Version="4.5.2" />
-    <PackageReference Include="Google.Protobuf" Version="3.11.3" />
-    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.2.0" />
+    <PackageReference Include="Microsoft.NETCore.Targets" Version="5.0.0" />
```
Comment (yuslepukhin) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests/Microsoft.ML.OnnxRuntime.Tests.csproj:63]: >crosoft.NETCore.Targets" Version="5.0.0" / [](start = 33, length = 42)

will revert

File: csharp/tools/Microsoft.ML.OnnxRuntime.PerfTool/Microsoft.ML.OnnxRuntime.PerfTool.csproj [hunk 2]
```diff
@@ -57,8 +57,8 @@
 
   <ItemGroup>
     <ProjectReference Include="$(OnnxRuntimeCSharpRoot)\src\Microsoft.ML.OnnxRuntime\Microsoft.ML.OnnxRuntime.csproj" />
-    <PackageReference Include="CommandLineParser" Version="2.4.3" />
-    <PackageReference Include="Google.Protobuf" Version="3.13.0" />
+    <PackageReference Include="CommandLineParser" Version="2.9.0-preview1" />
+    <PackageReference Include="Google.Protobuf" Version="3.14.0" />
```
Comment (yuslepukhin) [inline csharp/tools/Microsoft.ML.OnnxRuntime.PerfTool/Microsoft.ML.OnnxRuntime.PerfTool.csproj:61]: Will revert

File: tools/ci_build/github/Doxyfile_csharp.cfg [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+GENERATE_TODOLIST      = YES
+GENERATE_TESTLIST      = YES
+GENERATE_BUGLIST       = YES
+GENERATE_DEPRECATEDLIST= YES
+ENABLED_SECTIONS       =
+MAX_INITIALIZER_LINES  = 30
+SHOW_USED_FILES        = NO
+SHOW_FILES             = YES
+SHOW_NAMESPACES        = YES
+FILE_VERSION_FILTER    = "git -C c:\doxygen_src log  -n 1 --format=%h -- afile"
```
Comment (yuslepukhin) [inline tools/ci_build/github/Doxyfile_csharp.cfg:104]: >c:\doxygen_src [](start = 33, length = 14)

Need to fix this

File: csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.cs [hunk 4]
```diff
@@ -82,6 +88,7 @@ internal static IntPtr Handle
         /// Returns an instance of OrtEnv
         /// It returns the same instance on every call - `OrtEnv` is singleton
         /// </summary>
+        /// <returns>Returns a singleton instance or OrtEnv that represents native OrtEnv object</returns>
```
Comment (pranavsharma) [inline csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.cs:91]:
```suggestion
        /// <returns>Returns a singleton instance of OrtEnv that represents native OrtEnv object</returns>
```

File: csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.cs [hunk 5]
```diff
@@ -15,13 +15,19 @@ internal struct GlobalOptions  //Options are currently not accessible to user
...
-        Verbose = 0,
-        Info = 1,
-        Warning = 2,
-        Error = 3,
-        Fatal = 4
+        Verbose = 0, // Everything
+        Info = 1,    // Informational
+        Warning = 2, // Warnings
+        Error = 3,   // Errors
+        Fatal = 4    // This is it
```
Comment (pranavsharma) [inline csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.cs:30]:
```suggestion
        Fatal = 4    // Results in the termination of the application.
```

File: csharp/src/Microsoft.ML.OnnxRuntime/SessionOptions.cs [hunk 6]
```diff
@@ -279,11 +299,16 @@ internal IntPtr Handle
...
+        /// <summary>
+        /// Overrides SafeHandle.IsInvalid
+        /// </summary>
+        /// <value>returns true if handle is equal to Zero</value>
         public override bool IsInvalid { get { return handle == IntPtr.Zero; } }
 
         /// <summary>
         /// Enables the use of the memory allocation patterns in the first Run() call for subsequent runs. Default = true.
         /// </summary>
+        /// <value>returns enableMemoryPatter flag value</value>
```
Comment (pranavsharma) [inline csharp/src/Microsoft.ML.OnnxRuntime/SessionOptions.cs:311]:
```suggestion
        /// <value>returns enableMemoryPattern flag value</value>
```

File: csharp/src/Microsoft.ML.OnnxRuntime/SessionOptions.cs [hunk 7]
```diff
@@ -2,14 +2,15 @@
...
+using System.Runtime.InteropServices;
 using System.Text;
 using System.Runtime.InteropServices;
 using System.IO;
 
 namespace Microsoft.ML.OnnxRuntime
 {
     /// <summary>
-    /// TODO Add documentation about which optimizations are enabled for each value.
+    /// Graph optimization level to use with SessionOptions
```
Comment (pranavsharma) [inline csharp/src/Microsoft.ML.OnnxRuntime/SessionOptions.cs:13]: Missed this earlier. Can point to https://github.com/microsoft/onnxruntime/blob/master/docs/ONNX_Runtime_Graph_Optimizations.md.

**Review by pranavsharma — APPROVED** [review]

**Review by pranavsharma — APPROVED** [review]

**Review by pranavsharma — APPROVED** [review]

File: csharp/src/Microsoft.ML.OnnxRuntime/OrtIoBinding.cs [hunk 8]
```diff
@@ -238,6 +242,11 @@ public void ClearBoundOutputs()
         }
 
         #region SafeHandle
+        /// <summary>
+        /// Overrides SafeHandle.ReleaseHandle() to properly dispose of
+        /// the native instance of OrtIoBidning
```
Comment (edgchen1) [inline csharp/src/Microsoft.ML.OnnxRuntime/OrtIoBinding.cs:247]: OrtIoBidning -> OrtIoBinding

File: csharp/src/Microsoft.ML.OnnxRuntime/OrtValue.cs [hunk 9]
```diff
@@ -117,14 +121,17 @@ out ortValueHandle
...
         /// <param name="value">Tensor object</param>
         /// <param name="memoryHandle">For all tensor types but string tensors we endeavor to use managed memory
-        ///  to avoid additional allocation and copy. This out parameter represents a chunk of pinned memory
+        ///  to avoid additional allocation and copy. This out parameter represents a chunk of pinned memory which will need
+        ///  to be disposed when no longer needed. The lifespan of memoryHandle should eclipse the lifespan of the corresponding
+        ///  OrtValue.
         /// </param>
         /// <param name="elementType">discovered tensor element type</param>
-        /// <returns></returns>
+        /// <returns>And instance of OrtValue constructed on top of the object</returns>
```
Comment (edgchen1) [inline csharp/src/Microsoft.ML.OnnxRuntime/OrtValue.cs:134]: And -> An

File: csharp/src/Microsoft.ML.OnnxRuntime/RunOptions.cs [hunk 10]
```diff
@@ -16,13 +18,19 @@ internal IntPtr Handle
             }
         }
 
-
+        /// <summary>
+        /// Default __ctor. Creates default RuntimeOptions
```
Comment (edgchen1) [inline csharp/src/Microsoft.ML.OnnxRuntime/RunOptions.cs:22]: __ctor -> ctor

File: csharp/src/Microsoft.ML.OnnxRuntime/SessionOptions.cs [hunk 11]
```diff
@@ -421,6 +453,7 @@ public OrtLoggingLevel LogSeverityLevel
         /// Log Verbosity Level for the session logs. Default = 0. Valid values are >=0.
         /// This takes into effect only when the LogSeverityLevel is set to ORT_LOGGING_LEVEL_VERBOSE.
         /// </summary>
+        /// <value>returns _logVerbosityLevel value</value>
```
Comment (edgchen1) [inline csharp/src/Microsoft.ML.OnnxRuntime/SessionOptions.cs:456]: nit: a lot of these <value> comments seem to not provide useful information, just implementation details

File: csharp/src/Microsoft.ML.OnnxRuntime/Tensors/Tensor.cs [hunk 12]
```diff
@@ -281,6 +323,10 @@ public class TensorBase
          }
 
         private readonly Type _primitiveType;
+        /// <summary>
+        /// Constructs TensorBae
```
Comment (edgchen1) [inline csharp/src/Microsoft.ML.OnnxRuntime/Tensors/Tensor.cs:327]: TensorBae -> TensorBase

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/Microsoft.ML.OnnxRuntime.Tests.csproj [hunk 13]
```diff
@@ -65,7 +65,7 @@
     <PackageReference Include="Google.Protobuf" Version="3.11.3" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.2.0" />
     <PackageReference Include="xunit" Version="2.4.1" />
-    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1" />
```
Comment (edgchen1) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests/Microsoft.ML.OnnxRuntime.Tests.csproj:68]: nit: general comment, trailing whitespace

**Pull request merged**